meeteval/viz/visualize.js

```diff
@@ -23,6 +23,169 @@ var colormaps = {
+
+/**
+ * Call `fn` earliest after `delay` ms, but only if no other call is made in the meantime.
+ * If the function is called again within `delay` ms, the first call is canceled
+ * and the second call is postponed by `delay` ms.
+ *
+ * Example call: `call_delayed_throttled(this.setURL.bind(this), this.setURL, 200);`
+ */
+function call_delayed_throttled(fn, object, delay=5) {
+    if (object.timerId) {
+        clearTimeout(object.timerId);
+    }
+    object.timerId = setTimeout(() => {
+        fn()
+        object.timerId = undefined;
+    }, delay);
+}
+
+/**
+ * Interpolate between a and b with ratio c/d.
+ * If c <= 0, return a, if c >= d, then return b.
+ */
+function interpolate (a, b, c, d) {
+    if (c <= 0 || !b) {
+        // For speedup and numeric
+        return a
+    } else if (c >= d) {
+        // For speedup and numeric
+        return b
+    } else {
+        // Interpolate such that the length ratio between c and c+1 is constant.
+        const length_a = a[1] - a[0];
+        const length_b = b[1] - b[0];
+        const log_length_a = Math.log(length_a);
+        const log_length_b = Math.log(length_b);
+
+        const length = Math.exp(log_length_b  + ((log_length_a - log_length_b) * (d-c) / d));
+
+        let ratio = (length - length_b) / (length_a - length_b)
+
+        if (isNaN(ratio)) return a;
+
+        ratio = 1 - Math.max(Math.min(ratio, 1), 0);
+        let ratio_2 = 1 - ratio
+        return [a[0] * ratio_2 + b[0] * ratio, a[1] * ratio_2 + b[1] * ratio]
+    }
+}
+
+/**
+ * Checks whether two ranges `a` and `b` are similar by relative tolerance `tolerance`.
+ */
+function similar_range(a, b, tolerance=0.00001){
+    // Note: a == [0, 1] will always return false, because arrays are compared by reference, not values
+    if (JSON.stringify(a) === JSON.stringify(b)) {
+        return true
+    }
+
+    if (!a || a.some(isNaN) || a.some(isNaN))
+        return false
+
+    const delta = Math.min(a[1] - a[0], b[1] - b[0]) * tolerance
+
+    if (a.length !== b.length) {
+        return false;
+    }
+    for (var i = 0; i < a.length; i++) {
+        if (Math.abs(a[i] - b[i]) > delta) {
+            return false; // Return false if any elements are different
+        }
+    }
+    return true
+}
+
+/**
+ * Ensures that every entry (index i) lies within its parent (index i - 1).
+ *
+ * Keeps the viewArea at `anchor` fixed.
+ *
+ * Returns an array of booleans indicating which viewAreas were changed.
+ */
+function adjustViewAreas(viewAreas, anchor, viewArea) {
+    const dirty = viewAreas.map(() => false);
+
+    const oldViewAreas = [...viewAreas];
+
+    if (viewArea) {
+        viewAreas[anchor] = viewArea;
+        dirty[anchor] = true;
+    }
+
+    // Update children of anchor (j > i)
+    for (let j = anchor + 1; j < viewAreas.length; j++) {
+        let v = viewAreas[j];
+        const parent = viewAreas[j - 1];
+        if (parent[1] - parent[0] < v[1] - v[0] || similar_range(oldViewAreas[j - 1], v)) {
```

Comment:
Should it be marked as dirty, when the values ares the same? A draw wouldn't be necessary.

Reply:
Solved, now only marked as dirty when `!similar_rate(viewAreas[j], v)`